src/reconcile.rs

```diff
@@ -285,6 +298,336 @@ where
+            + HasStatus
+            + Resource<DynamicType = ()>
+            + kube::CustomResourceExt,
+        <T as HasStatus>::Status: HasCurrentCommand + Debug + Default + Serialize,
+        C: Command + CanBeRolling + DeserializeOwned + HasRoles,
+        <C as kube::Resource>::DynamicType: Default,
+    {
+        // set start time in command once
+        if command.start_time().is_none() {
+            let patch = command.start_patch();
+            self.client.merge_patch_status(command, &patch).await?;
+        }
+
+        self.client
+            .merge_patch_status(
+                &self.resource,
+                &self
+                    .resource
+                    .cluster_execution_status_patch(&ClusterExecutionStatus::Running),
+            )
+            .await?;
+
+        clear_current_command(&self.client, &mut self.resource).await?;
+
+        let patch = command.finish_patch();
+        self.client.merge_patch_status(command, &patch).await?;
+
+        Ok(ReconcileFunctionAction::Done)
+    }
+
+    /// Create a LabelSelector with matching expressions ("In") containing:
+    /// - Application name
+    /// - Component (Role)
+    /// - Instance (cluster name)
+    fn get_role_selector(&self, role: &str) -> LabelSelector
+    where
+        T: HasApplication,
+    {
+        let application_match = LabelSelectorRequirement {
+            key: APP_NAME_LABEL.to_string(),
+            operator: "In".to_string(),
+            values: Some(vec![
+                <T as HasApplication>::get_application_name().to_string()
+            ]),
+        };
+        let role_match = LabelSelectorRequirement {
+            key: APP_COMPONENT_LABEL.to_string(),
+            operator: "In".to_string(),
+            values: Some(vec![role.to_string()]),
+        };
+
+        let instance_match = LabelSelectorRequirement {
+            key: APP_INSTANCE_LABEL.to_string(),
+            operator: "In".to_string(),
+            values: Some(vec![self.resource.name()]),
+        };
+
+        let result = LabelSelector {
+            match_expressions: Some(vec![application_match, role_match, instance_match]),
+            match_labels: Default::default(),
+        };
+        trace!("Created LabelSelector: [{:?}]", result);
+        result
+    }
+
+    /// This reconcile function returns an existing executing command or the next command
+    /// in the queue (with the oldest timestamp).
+    pub async fn retrieve_current_command(&mut self) -> OperatorResult<Option<CommandRef>>
+    where
+        T: Clone
+            + Debug
+            + DeserializeOwned
+            + Resource
+            + CustomResourceExt
+            + HasCommands
+            + HasStatus
+            + Send
+            + Sync
+            + 'static,
+        <T as Resource>::DynamicType: Default,
+        <T as HasStatus>::Status: HasCurrentCommand + Debug + Default + Serialize,
+    {
+        let current_command_ref = crate::command::current_command(
```

Comment:
I think we could make this more testable if we retrieved the status and list of commandrefs here and passed those into `current_command` instead of a client and a resource..

That would mean that we needlessly retrieve all commands even if a command is set in the status on every reconciliation though.

Reply:
We decided on doing that after the hackathon.